<a href="https://colab.research.google.com/github/k-rajmani2k/Generative-AI_Chatbot/blob/main/Financial_Analysis_Experience.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd

Manual extraction:

For each company, find the 10-K filings for the last three fiscal years.
Extract the following financial figures: Total Revenue, Net Income, Total Assets, Total Liabilities, and Cash Flow from Operating Activities.

In [ ]:
df = pd.read_csv('/content/Combined_Financial.csv')

In [ ]:
df

Company  Year  Total Revenue (in millions)  Net Income (in millions)  \
0      Tesla  2023                        96773                     14974   
1      Tesla  2022                        81462                     12587   
2      Tesla  2021                        53823                      5644   
3      Apple  2024                       391035                     93736   
4      Apple  2023                       383285                     96995   
5      Apple  2022                       394328                     99803   
6  Microsoft  2024                       245122                     88136   
7  Microsoft  2023                       211915                     72361   
8  Microsoft  2022                       198270                     72738   

   Gross Profit (in millions)  \
0                     17660.0   
1                     20853.0   
2                     13606.0   
3                    180683.0   
4                    169148.0   
5                    170782.0   
6                         NaN   
7                         NaN   
8                         NaN   

   Cash Flow from Operating Activities (in millions)  
0                                              21772  
1                                              18152  
2                                              10816  
3                                             118254  
4                                             110543  
5                                             122151  
6                                             118548  
7                                              87582  
8                                              89035

In [ ]:
# df.rename(columns={'Total Revenue (in millions)': 'Total Revenue'}, inplace=True)
# df.rename(columns={'Net Income (in millions)': 'Net Income'}, inplace=True)

In [ ]:
df.rename(columns={
    'Total Revenue (in millions)': 'Total Revenue',
    'Net Income (in millions)': 'Net Income'
}, inplace=True)


In [ ]:
df

Company  Year  Total Revenue  Net Income  Gross Profit (in millions)  \
0      Tesla  2023          96773       14974                     17660.0   
1      Tesla  2022          81462       12587                     20853.0   
2      Tesla  2021          53823        5644                     13606.0   
3      Apple  2024         391035       93736                    180683.0   
4      Apple  2023         383285       96995                    169148.0   
5      Apple  2022         394328       99803                    170782.0   
6  Microsoft  2024         245122       88136                         NaN   
7  Microsoft  2023         211915       72361                         NaN   
8  Microsoft  2022         198270       72738                         NaN   

   Cash Flow from Operating Activities (in millions)  
0                                              21772  
1                                              18152  
2                                              10816  
3                                             118254  
4                                             110543  
5                                             122151  
6                                             118548  
7                                              87582  
8                                              89035

Analyzing trends with pandas:

Use pandas to calculate year-over-year changes for each financial metric. You can do this by creating new columns in your DataFrame that represent the percentage change from one year to the next.

In [ ]:
df['Revenue Growth (%)'] = df.groupby(['Company'])['Total Revenue'].pct_change() * 100
df['Net Income Growth (%)'] = df.groupby(['Company'])['Net Income'].pct_change() * 100


In [ ]:
df

Company  Year  Total Revenue  Net Income  Gross Profit (in millions)  \
0      Tesla  2023          96773       14974                     17660.0   
1      Tesla  2022          81462       12587                     20853.0   
2      Tesla  2021          53823        5644                     13606.0   
3      Apple  2024         391035       93736                    180683.0   
4      Apple  2023         383285       96995                    169148.0   
5      Apple  2022         394328       99803                    170782.0   
6  Microsoft  2024         245122       88136                         NaN   
7  Microsoft  2023         211915       72361                         NaN   
8  Microsoft  2022         198270       72738                         NaN   

   Cash Flow from Operating Activities (in millions)  Revenue Growth (%)  \
0                                              21772                 NaN   
1                                              18152          -15.821562   
2                                              10816          -33.928703   
3                                             118254                 NaN   
4                                             110543           -1.981920   
5                                             122151            2.881146   
6                                             118548                 NaN   
7                                              87582          -13.547132   
8                                              89035           -6.438902   

   Net Income Growth (%)  
0                    NaN  
1             -15.940964  
2             -55.160086  
3                    NaN  
4               3.476786  
5               2.894995  
6                    NaN  
7             -17.898475  
8               0.520999

Summarizing findings:

Conclude your analysis by summarizing your findings directly in the notebook. Use markdown cells to add narrative explanations of your analysis, discussing the trends and changes in financial metrics you've identified.

In [ ]:
df.describe()

Year  Total Revenue    Net Income  Gross Profit (in millions)  \
count     9.000000       9.000000      9.000000                    6.000000   
mean   2022.666667  228445.888889  61886.000000                95455.333333   
std       1.000000  136392.979105  39363.064871                85656.764346   
min    2021.000000   53823.000000   5644.000000                13606.000000   
25%    2022.000000   96773.000000  14974.000000                18458.250000   
50%    2023.000000  211915.000000  72738.000000                95000.500000   
75%    2023.000000  383285.000000  93736.000000               170373.500000   
max    2024.000000  394328.000000  99803.000000               180683.000000   

       Cash Flow from Operating Activities (in millions)  Revenue Growth (%)  \
count                                           9.000000            6.000000   
mean                                        77428.111111          -11.472845   
std                                         47091.116268           13.030233   
min                                         10816.000000          -33.928703   
25%                                         21772.000000          -15.252954   
50%                                         89035.000000           -9.993017   
75%                                        118254.000000           -3.096165   
max                                        122151.000000            2.881146   

       Net Income Growth (%)  
count               6.000000  
mean              -13.684458  
std                22.424357  
min               -55.160086  
25%               -17.409097  
50%                -7.709983  
75%                 2.301496  
max                 3.476786

In [ ]:
df

Company  Year  Total Revenue  Net Income  Gross Profit (in millions)  \
0      Tesla  2023          96773       14974                     17660.0   
1      Tesla  2022          81462       12587                     20853.0   
2      Tesla  2021          53823        5644                     13606.0   
3      Apple  2024         391035       93736                    180683.0   
4      Apple  2023         383285       96995                    169148.0   
5      Apple  2022         394328       99803                    170782.0   
6  Microsoft  2024         245122       88136                         NaN   
7  Microsoft  2023         211915       72361                         NaN   
8  Microsoft  2022         198270       72738                         NaN   

   Cash Flow from Operating Activities (in millions)  Revenue Growth (%)  \
0                                              21772                 NaN   
1                                              18152          -15.821562   
2                                              10816          -33.928703   
3                                             118254                 NaN   
4                                             110543           -1.981920   
5                                             122151            2.881146   
6                                             118548                 NaN   
7                                              87582          -13.547132   
8                                              89035           -6.438902   

   Net Income Growth (%)  
0                    NaN  
1             -15.940964  
2             -55.160086  
3                    NaN  
4               3.476786  
5               2.894995  
6                    NaN  
7             -17.898475  
8               0.520999

In [ ]:
# Fill NA values that result from pct_change calculations with 0 or an appropriate value
df.fillna(0, inplace=True)

# Display the dataframe to verify the calculations
print(df)

# Optionally, you could summarize these findings for each company
summary = df.groupby('Company').agg({
    'Revenue Growth (%)': 'mean',
    'Net Income Growth (%)': 'mean'
}).reset_index()

print("\nYear-over-Year Average Growth Rates (%):")
print(summary)

     Company  Year  Total Revenue  Net Income  Gross Profit (in millions)  \
0      Tesla  2023          96773       14974                     17660.0   
1      Tesla  2022          81462       12587                     20853.0   
2      Tesla  2021          53823        5644                     13606.0   
3      Apple  2024         391035       93736                    180683.0   
4      Apple  2023         383285       96995                    169148.0   
5      Apple  2022         394328       99803                    170782.0   
6  Microsoft  2024         245122       88136                         0.0   
7  Microsoft  2023         211915       72361                         0.0   
8  Microsoft  2022         198270       72738                         0.0   

   Cash Flow from Operating Activities (in millions)  Revenue Growth (%)  \
0                                              21772            0.000000   
1                                              18152          -15.821562   
2

In [ ]:
summary

Company  Revenue Growth (%)  Net Income Growth (%)
0      Apple            0.299742               2.123927
1  Microsoft           -6.662011              -5.792492
2      Tesla          -16.583422             -23.700350

### **Financial Analysis Summary**

#### **Overview**
This analysis evaluates the financial performance of Tesla, Apple, and Microsoft over the past three fiscal years using key metrics: Total Revenue, Net Income, Gross Profit, Cash Flow from Operating Activities, Revenue Growth (%), and Net Income Growth (%). The following insights were derived from the data:

---

#### **Tesla**
1. **Revenue and Profit Trends**:
   - Tesla's **Total Revenue** showed consistent growth, with a notable increase in 2022 (+15.82% YoY). However, **Net Income** dropped significantly in 2022 (-55.16%) due to rising costs or operational challenges.
   - The **Gross Profit** in 2023 decreased compared to 2022, indicating potential cost management issues.

2. **Cash Flow from Operating Activities**:
   - Tesla's cash flow consistently improved, reflecting operational strength and efficiency over time.

3. **Growth Rates**:
   - Both Revenue Growth and Net Income Growth declined between 2022 and 2023, suggesting Tesla faced challenges in scaling profitability during these years.

---

#### **Apple**
1. **Revenue and Profit Trends**:
   - Apple maintained steady revenue levels, with minor fluctuations between years. Revenue Growth was relatively flat, hovering between -1.98% (2023) and +2.88% (2022).
   - Net Income showed slight positive growth in 2023 (+3.48%) and 2022 (+2.89%), indicating stable profit margins.

2. **Gross Profit and Cash Flow**:
   - Apple's **Gross Profit** remained high, driven by a healthy balance of product and service revenues.
   - Strong **Cash Flow from Operating Activities** reflects the company's efficient revenue conversion.

3. **Stability**:
   - Apple has a consistent financial performance, showcasing its resilience and maturity as a leading technology firm.

---

#### **Microsoft**
1. **Revenue and Profit Trends**:
   - Microsoft's revenue grew consistently until 2023 but saw a decline in 2024 (-13.55% YoY). Net Income followed a similar trend with a drop of -17.90% in 2024.
   - The decline in growth rates may stem from market saturation or increased competition in its product lines.

2. **Cash Flow from Operating Activities**:
   - Despite declining revenue and income, Microsoft's operating cash flow grew, reflecting robust operational efficiency.

3. **Growth Challenges**:
   - Revenue and Net Income Growth in 2024 suggest Microsoft faces challenges in maintaining prior growth trajectories, warranting a focus on cost management and product innovation.

---

#### **Comparative Analysis**
- **Revenue Trends**: Apple remains the largest revenue generator, followed by Microsoft and Tesla. However, Tesla's growth rates were more dynamic, signaling rapid expansion.
- **Profitability**: Apple consistently maintains high Net Income, indicating robust margins and operational stability. Tesla shows greater variability in profits, while Microsoft experienced a notable drop in 2024.
- **Cash Flow**: All three companies maintain positive operating cash flow, with Apple's and Microsoft's efficiency standing out.
- **Growth Rates**: Tesla experienced the most volatility, Microsoft faced challenges in sustaining growth in 2024, and Apple remained relatively steady.

---

#### **Key Takeaways**
1. **Tesla**: Needs to address cost management and improve profitability as revenue grows. A focus on stabilizing net income is crucial.
2. **Apple**: Demonstrates stability and operational excellence, making it a benchmark for mature companies.
3. **Microsoft**: Faces growth challenges and should prioritize innovation and market expansion to maintain competitive edge.

These insights provide a strong foundation for building an AI-powered financial chatbot that can guide users in interpreting financial performance and trends for major companies.

Task 2: Developing an AI-powered financial chatbot

1

2

3

4

5

6

7

8

9

10
Example answer
Great work! Take a look at the example answer below to see how a professional would have attempted this task. Think about what you did well and how you can improve.

Let's consider a simplified example answer for developing a financial chatbot prototype. This example aims to guide you through the process and principles involved but remember that there's room for creativity and innovation based on your skills and understanding.

Chatbot example: Basic financial inquiry bot
Objective: Develop a chatbot that responds to predefined queries about a company's financial performance, specifically focusing on net income and revenue growth.

1. Rule-based logic implementation
Python code snippet



In [ ]:
def financial_chatbot(query):
    responses = {
        "net income": "The net income for the fiscal year 2021 was $5 billion, a 10% increase from 2020.",
        "revenue growth": "The revenue grew by 8% from 2020 to 2021, reaching $22 billion."
    }
    return responses.get(query.lower(), "I'm not sure how to answer that. Can you try a different question?")

In [ ]:
import pandas as pd

# Load data from CSV file
def load_data(file_path):
    try:
        df = pd.read_csv(file_path)
        return df
    except Exception as e:
        print(f"Error loading data: {e}")
        return None

# Financial chatbot function
def financial_chatbot(query, data):
    # Check if data is loaded
    if data is None or data.empty:
        return "Financial data is not available. Please ensure the CSV file is loaded correctly."

    query = query.lower()
    # Handle predefined queries
    if "net income" in query:
        return net_income_query(data)
    elif "total revenue" in query:
        return total_revenue_query(data)
    elif "revenue growth" in query:
        return revenue_growth_query(data)
    elif "cash flow" in query:
        return cash_flow_query(data)
    else:
        return "I'm sorry, I can only answer predefined financial queries. Please try again."

# Helper functions for specific queries
def net_income_query(data):
    company = get_company_from_query()
    result = []
    for index, row in data.iterrows():
        result.append(f"{row['Company']} in {row['Year']} net incione was X.")


In [ ]:
financial_chatbot("net income")

TypeError: financial_chatbot() missing 1 required positional argument: 'data'

Explanation: This simple Python function uses a dictionary to map user queries to predefined responses. It demonstrates the principle of rule-based logic and error handling by providing a default response when the query does not match.

2. Data structuring and retrieval
Data structuring: Assume we have a CSV file named financial_data.csv that stores our company's financial information. For this example, the chatbot's responses are hardcoded, but in practice, you would retrieve this information from your structured data source.

Retrieval method: While our example does not directly fetch data from a file, in a more advanced scenario, you could use pandas to read the CSV and query the DataFrame for specific information based on the user's input.

3. Communicating financial insights
Dialogue design: The chatbot's responses are crafted to be clear and informative, summarizing complex financial data into easy-to-understand insights. For a more interactive dialogue, you might expand the function to suggest related queries or follow-up for more detail.


Example interaction with the chatbot
User query: "What was the net income?"

Chatbot response: "The net income for the fiscal year 2021 was $5 billion, a 10% increase from 2020."

User query: "Tell me about revenue growth."

Chatbot response: "Revenue grew by 8% from 2020 to 2021, reaching $22 billion."

Conclusion and variation:
This example demonstrates the core principles of developing a financial chatbot using rule-based logic. Depending on your depth of understanding and technical skills, there are numerous ways to enhance and expand this prototype. You might integrate more complex NLP features for natural language understanding, connect to real-time financial databases for dynamic data retrieval, or implement machine learning for more nuanced interaction and response generation.

Key takeaway: The principles of chatbot development (rule-based logic, data integration, and clear communication) apply regardless of the complexity of your chatbot. This example provides a foundation, but the possibilities are vast, encouraging you to explore and innovate beyond this basic framework.


